File: src/bin/lazer_probe.rs
//! Live verification of Pyth Lazer pre-positioning (run on the box, where
//! PYTH_LAZER_TOKEN lives). Streams the volatile majors, scans the marginfi
//! watch-set once, then each interval recomputes health with Lazer prices
//! blended over the on-chain baseline and prints the nearest-to-liquidation
//! accounts + the Lazer-vs-on-chain price delta per major. Confirms (a) the
//! feed is live, (b) the mint→feed mapping resolves banks, and (c) Lazer leads
//! the on-chain oracle (nonzero delta = the pre-positioning edge).
//!
//! Usage: HELIUS_RPC=<url> PYTH_LAZER_TOKEN=<token> [INTERVAL_MS=2000]
//!        cargo run --release --bin lazer_probe

use arb_engine::lazer;
use arb_engine::liquidation::{self as liq, Bank, BankMap, MarginfiAccount, PriceMap};
use arb_engine::pyth;
use solana_pubkey::Pubkey;
use std::collections::{HashMap, HashSet};
use std::time::Duration;

const MARGINFI_PROGRAM: &str = "MFv2hWf31Z9kbCa1snEPYctwafyhdvnV7FZnsebVacA";
const MARGINFI_GROUP: &str = "4qp6Fx6tnZkY5Wropq9wUYgtFxXKwE6viZxFHg3rdAG8";

fn rpc(endpoint: &str, body: serde_json::Value) -> Option<serde_json::Value> {
    for attempt in 0..4 {
        if let Ok(resp) = ureq::post(endpoint).send_json(body.clone()) {
            if let Ok(v) = resp.into_json::<serde_json::Value>() { return Some(v); }
        }
        std::thread::sleep(Duration::from_millis(400 << attempt));
    }
    None
}
fn b64(d: &serde_json::Value) -> Option<Vec<u8>> {
    use base64::Engine;
    base64::engine::general_purpose::STANDARD.decode(d.get(0)?.as_str()?).ok()
}
fn get_multiple(endpoint: &str, keys: &[Pubkey]) -> HashMap<Pubkey, Vec<u8>> {
    let mut out = HashMap::new();
    for chunk in keys.chunks(100) {
        let strs: Vec<String> = chunk.iter().map(|k| k.to_string()).collect();
        let Some(v) = rpc(endpoint, serde_json::json!({"jsonrpc":"2.0","id":1,"method":"getMultipleAccounts",
            "params":[strs, {"encoding":"base64"}]})) else { continue };
        for (i, acc) in v["result"]["value"].as_array().into_iter().flatten().enumerate() {
            if let Some(b) = acc.get("data").and_then(b64) { out.insert(chunk[i], b); }
        }
    }
    out
}

fn main() {
    let _ = dotenvy::dotenv();
    let _ = rustls::crypto::ring::default_provider().install_default();
    let endpoint = std::env::var("HELIUS_RPC").or_else(|_| std::env::var("RPC_HTTP")).expect("HELIUS_RPC");
    let token = std::env::var("PYTH_LAZER_TOKEN").expect("PYTH_LAZER_TOKEN (lives on the box)");
    let interval = Duration::from_millis(std::env::var("INTERVAL_MS").ok().and_then(|s| s.parse().ok()).unwrap_or(2000));

    let table = pyth::new_table();
    lazer::spawn_lazer_thread(token, lazer::arm_feed_ids(), table.clone());
    eprintln!("[lazer] subscribed to majors; scanning marginfi group …");

    // Scan borrowers + banks + on-chain oracle prices once (baseline).
    let resp = rpc(&endpoint, serde_json::json!({"jsonrpc":"2.0","id":1,"method":"getProgramAccounts",
        "params":[MARGINFI_PROGRAM, {"encoding":"base64","dataSlice":{"offset":0,"length":1736},
            "filters":[{"dataSize":liq::MA_SIZE},{"memcmp":{"offset":8,"bytes":MARGINFI_GROUP}}]}]}));
    let entries = resp.as_ref().and_then(|v| v["result"].as_array()).cloned().unwrap_or_default();
    let accts: Vec<(Pubkey, MarginfiAccount)> = entries.iter().filter_map(|e| {
        Some((e["pubkey"].as_str()?.parse().ok()?, MarginfiAccount::decode(&b64(&e["account"]["data"])?)?))
    }).filter(|(_, a): &(Pubkey, MarginfiAccount)| a.balances.iter().any(|b| b.liability_shares > 0.0)).collect();
    let bank_pks: Vec<Pubkey> = accts.iter().flat_map(|(_, a)| a.balances.iter().map(|b| b.bank_pk)).collect::<HashSet<_>>().into_iter().collect();
    let mut banks: BankMap = HashMap::new();
    let mut oracle_of: HashMap<Pubkey, Pubkey> = HashMap::new();
    for (pk, raw) in &get_multiple(&endpoint, &bank_pks) {
        if let Some(bk) = Bank::decode(raw) { oracle_of.insert(*pk, bk.oracle_key); banks.insert(*pk, bk); }
    }
    let oracle_pks: Vec<Pubkey> = oracle_of.values().copied().collect::<HashSet<_>>().into_iter().collect();
    let mut on_chain: PriceMap = HashMap::new();
    for (pk, raw) in &get_multiple(&endpoint, &oracle_pks) {
        if let Some(usd) = liq::decode_oracle_price(raw) {
            for (bk, oc) in &oracle_of { if oc == pk { on_chain.insert(*bk, usd); } }
        }
    }
    let map = lazer::mint_feed_map();
    eprintln!("[lazer] {} borrowers, {} banks, {} on-chain-priced", accts.len(), banks.len(), on_chain.len());

    loop {
        std::thread::sleep(interval);
        if pyth::get(&table, lazer::LAZER_SOL).is_none() {
            eprintln!("[lazer] waiting for first tick …");
            continue;
        }
        let (blended, led) = lazer::blend(&banks, &on_chain, &table, &map);
        // Lazer-vs-on-chain delta on SOL (the leading-edge proof).
        let sol_delta = banks.iter().find(|(_, b)| map.get(&b.mint) == Some(&lazer::LAZER_SOL))
            .and_then(|(pk, _)| Some((on_chain.get(pk)?, blended.get(pk)?)))
            .map(|(oc, lz)| format!("SOL on-chain ${oc:.2} → Lazer ${lz:.2} (Δ{:+.2})", lz - oc))
            .unwrap_or_else(|| "no SOL bank".into());

        // Nearest-to-liquidation by Lazer-blended health.
        let mut ranked: Vec<(Pubkey, f64, f64)> = accts.iter().filter_map(|(pk, a)| {
            let r = liq::maintenance_health(a, &banks, &blended);
            (r.missing == 0 && r.health.weighted_assets >= 100.0).then_some((*pk, r.health.ratio(), r.health.weighted_assets))
        }).collect();
        ranked.sort_by(|a, b| b.1.total_cmp(&a.1));
        println!("\n[{}] {}  ({} banks Lazer-led)", lazer::status(&table), sol_delta, led);
        for (pk, ratio, assets) in ranked.iter().take(5) {
            println!("  {}  ratio {:.4}  collateral ${:.0}{}",
                &pk.to_string()[..8], ratio, assets, if *ratio >= 1.0 { "  ← LIQUIDATABLE (Lazer)" } else { "" });
        }
    }
}

File: src/bin/lazerProbe.ts
// Port of src/bin/lazer_probe.rs
//
// Live verification of Pyth Lazer pre-positioning (run on the box, where
// PYTH_LAZER_TOKEN lives). Streams the volatile majors, scans the marginfi
// watch-set once, then each interval recomputes health with Lazer prices
// blended over the on-chain baseline and prints the nearest-to-liquidation
// accounts + the Lazer-vs-on-chain price delta per major. Confirms (a) the
// feed is live, (b) the mint→feed mapping resolves banks, and (c) Lazer leads
// the on-chain oracle (nonzero delta = the pre-positioning edge).
//
// Usage: HELIUS_RPC=<url> PYTH_LAZER_TOKEN=<token> [INTERVAL_MS=2000]
//        tsx src/bin/lazerProbe.ts

import 'dotenv/config';
import { PublicKey } from '@solana/web3.js';
import { armFeedIds, blend, LAZER_SOL, mintFeedMap, spawnLazerThread, status } from '../lib/lazer.js';
import {
  decodeBank,
  decodeMarginfiAccount,
  decodeOraclePrice,
  MA_SIZE,
  maintenanceHealth,
  healthRatio,
  type Bank,
  type BankMap,
  type MarginfiAccount,
  type PriceMap,
} from '../lib/liquidation.js';
import { get, newTable } from '../lib/pyth.js';

const MARGINFI_PROGRAM = 'MFv2hWf31Z9kbCa1snEPYctwafyhdvnV7FZnsebVacA';
const MARGINFI_GROUP = '4qp6Fx6tnZkY5Wropq9wUYgtFxXKwE6viZxFHg3rdAG8';

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function rpc(endpoint: string, body: unknown): Promise<any | undefined> {
  for (let attempt = 0; attempt < 4; attempt++) {
    try {
      const resp = await fetch(endpoint, {
        method: 'POST',
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify(body),
      });
      return await resp.json();
    } catch {
      // fall through to retry
    }
    await sleep(400 << attempt);
  }
  return undefined;
}

function b64(d: any): Buffer | undefined {
  const s = d?.[0];
  if (typeof s !== 'string') return undefined;
  try {
    return Buffer.from(s, 'base64');
  } catch {
    return undefined;
  }
}

async function getMultiple(endpoint: string, keys: PublicKey[]): Promise<Map<string, Buffer>> {
  const out = new Map<string, Buffer>();
  for (let i = 0; i < keys.length; i += 100) {
    const chunk = keys.slice(i, i + 100);
    const strs = chunk.map((k) => k.toBase58());
    const v = await rpc(endpoint, {
      jsonrpc: '2.0',
      id: 1,
      method: 'getMultipleAccounts',
      params: [strs, { encoding: 'base64' }],
    });
    const arr: any[] = Array.isArray(v?.result?.value) ? v.result.value : [];
    arr.forEach((acc, idx) => {
      const data = b64(acc?.data);
      if (data !== undefined) out.set(chunk[idx].toBase58(), data);
    });
  }
  return out;
}

async function main(): Promise<void> {
  const endpoint = process.env.HELIUS_RPC ?? process.env.RPC_HTTP;
  if (endpoint === undefined) throw new Error('HELIUS_RPC');
  const token = process.env.PYTH_LAZER_TOKEN;
  if (token === undefined) throw new Error('PYTH_LAZER_TOKEN (lives on the box)');
  const interval = Number.parseInt(process.env.INTERVAL_MS ?? '', 10) || 2000;

  const table = newTable();
  spawnLazerThread(token, armFeedIds(), table);
  console.error('[lazer] subscribed to majors; scanning marginfi group …');

  // Scan borrowers + banks + on-chain oracle prices once (baseline).
  const resp = await rpc(endpoint, {
    jsonrpc: '2.0',
    id: 1,
    method: 'getProgramAccounts',
    params: [
      MARGINFI_PROGRAM,
      {
        encoding: 'base64',
        dataSlice: { offset: 0, length: 1736 },
        filters: [{ dataSize: MA_SIZE }, { memcmp: { offset: 8, bytes: MARGINFI_GROUP } }],
      },
    ],
  });
  const entries: any[] = Array.isArray(resp?.result) ? resp.result : [];
  const accts: [PublicKey, MarginfiAccount][] = [];
  for (const e of entries) {
    const pkStr = e?.pubkey;
    const data = b64(e?.account?.data);
    if (typeof pkStr !== 'string' || data === undefined) continue;
    const acct = decodeMarginfiAccount(data);
    if (acct === null) continue;
    if (!acct.balances.some((b) => b.liabilityShares > 0.0)) continue;
    accts.push([new PublicKey(pkStr), acct]);
  }
  const bankPksSet = new Set<string>();
  const bankPks: PublicKey[] = [];
  for (const [, a] of accts) {
    for (const b of a.balances) {
      const key = b.bankPk.toBase58();
      if (!bankPksSet.has(key)) {
        bankPksSet.add(key);
        bankPks.push(b.bankPk);
      }
    }
  }
  const banks: BankMap = new Map();
  const oracleOf = new Map<string, PublicKey>();
  const bankRaw = await getMultiple(endpoint, bankPks);
  for (const [pkStr, raw] of bankRaw) {
    const bk = decodeBank(raw);
    if (bk !== null) {
      oracleOf.set(pkStr, bk.oracleKey);
      banks.set(pkStr, bk);
    }
  }
  const oraclePksSet = new Set<string>();
  const oraclePks: PublicKey[] = [];
  for (const oc of oracleOf.values()) {
    const key = oc.toBase58();
    if (!oraclePksSet.has(key)) {
      oraclePksSet.add(key);
      oraclePks.push(oc);
    }
  }
  const onChain: PriceMap = new Map();
  const oracleRaw = await getMultiple(endpoint, oraclePks);
  for (const [pkStr, raw] of oracleRaw) {
    const usd = decodeOraclePrice(raw);
    if (usd !== null) {
      for (const [bk, oc] of oracleOf) {
        if (oc.toBase58() === pkStr) onChain.set(bk, usd);
      }
    }
  }
  const map = mintFeedMap();
  console.error(`[lazer] ${accts.length} borrowers, ${banks.size} banks, ${onChain.size} on-chain-priced`);

  for (;;) {
    await sleep(interval);
    if (get(table, LAZER_SOL) === undefined) {
      console.error('[lazer] waiting for first tick …');
      continue;
    }
    const [blended, led] = blend(banks, onChain, table, map);
    // Lazer-vs-on-chain delta on SOL (the leading-edge proof).
    let solDelta = 'no SOL bank';
    for (const [pk, b] of banks) {
      if (map.get(b.mint.toBase58()) === LAZER_SOL) {
        const oc = onChain.get(pk);
        const lz = blended.get(pk);
        if (oc !== undefined && lz !== undefined) {
          solDelta = `SOL on-chain $${oc.toFixed(2)} → Lazer $${lz.toFixed(2)} (Δ${lz - oc >= 0 ? '+' : ''}${(lz - oc).toFixed(2)})`;
        }
        break;
      }
    }

    // Nearest-to-liquidation by Lazer-blended health.
    const ranked: [PublicKey, number, number][] = [];
    for (const [pk, a] of accts) {
      const r = maintenanceHealth(a, banks, blended);
      if (r.missing === 0 && r.health.weightedAssets >= 100.0) {
        ranked.push([pk, healthRatio(r.health), r.health.weightedAssets]);
      }
    }
    ranked.sort((a, b) => b[1] - a[1]);
    console.log(`\n[${status(table)}] ${solDelta}  (${led} banks Lazer-led)`);
    for (const [pk, ratio, assets] of ranked.slice(0, 5)) {
      const pkStr = pk.toBase58();
      console.log(
        `  ${pkStr.slice(0, 8)}  ratio ${ratio.toFixed(4)}  collateral $${assets.toFixed(0)}${ratio >= 1.0 ? '  ← LIQUIDATABLE (Lazer)' : ''}`,
      );
    }
  }
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
